sdk/containerregistry/azure-containerregistry/azure/containerregistry/_authentication_policy.py

```diff
@@ -4,61 +4,99 @@
+    if option is False:
+        request.context["enforce_https"] = option
 
-    def get_token(self):
-        token_str = "{}:{}".format(self._user, self._password)
-        token_bytes = token_str.encode("ascii")
-        b64_bytes = b64encode(token_bytes)
-        return b64_bytes.decode("ascii")
+    enforce_https = request.context.get("enforce_https", True)
+    if enforce_https and not request.http_request.url.lower().startswith("https"):
+        raise ServiceRequestError(
+            "Bearer token authentication is not permitted for non-TLS protected (non-https) URLs."
+        )
 
 
-class ContainerRegistryUserCredentialPolicy(SansIOHTTPPolicy):
-    """HTTP pipeline policy to authenticate using ContainerRegistryUserCredential"""
+class ContainerRegistryChallengePolicy(HTTPPolicy):
+    """Authentication policy for ACR which accepts a challenge"""
 
-    def __init__(self, credential):
+    def __init__(self, credential, endpoint):
+        # type: (TokenCredential, str) -> None
+        super(ContainerRegistryChallengePolicy, self).__init__()
+        self._scopes = "https://management.core.windows.net/.default"
         self._credential = credential
+        self._token = None  # type: Optional[AccessToken]
+        self._exchange_client = ACRExchangeClient(endpoint, self._credential)
 
-    @staticmethod
-    def _update_headers(headers, token):
-        headers["Authorization"] = "Basic {}".format(token)
+    def _need_new_token(self):  # pylint: disable=no-self-use
+        # type: () -> bool
+        return True
 
     def on_request(self, request):
         # type: (PipelineRequest) -> None
-        self._update_headers(request.http_request.headers, self._credential.get_token())
-
-
-class ContainerRegistryCredentialPolicy(SansIOHTTPPolicy):
-    """Challenge based authentication policy for ACR. This policy is used for getting
-    the AAD Token, refresh token, and access token before performing a call to service.
-
-    :param credential: Azure Token Credential for authenticating with Azure
-    :type credential: TokenCredential
-    """
-
-    BEARER = "Bearer"
-    AUTHENTICATION_CHALLENGE_PARAMS_PATTERN = re.compile('(?:(\\w+)="([^""]*)")+')
-    WWW_AUTHENTICATE = "WWW-Authenticate"
-    SCOPE_PARAMETER = "scope"
-    SERVICE_PARAMETER = "service"
-    AUTHORIZATION = "Authorization"
-
-    def __init__(self, credential, url, pipeline):
-        self._credential = credential
-        self.url = url
-        self.pipeline = pipeline
+        """Called before the policy sends a request.
+        The base implementation authorizes the request with a bearer token.
+        :param ~azure.core.pipeline.PipelineRequest request: the request
+        """
+
+        if self._token is None or self._need_new_token():
+            self._token = self._credential.get_token(self._scopes)
+            try:
+                self._token = self._token.token
+            except AttributeError:
+                pass
```

Comment:
I'm confused about how this is supposed to work. `self._credential` here is a TokenCredential, so `self._token` is an AccessToken, meaning you won't get AttributeError, but line 67 will raise because you can't concatenate AccessToken to str. Then again, down in `on_challenge` you assign a string to `self._token`, so after handling a challenge you never enter this block and line 67 works. I suppose this functions to elicit a challenge from ACR, but why save the ARM token on `self._token`?

Reply:
When I add caching to the policy this functionality will significantly change. However, currently I am going to make the change to `pass` on the `on_request` because the original request does not need any `Authorization` header. In the future `self._token` will be an ACR Access Token.